rust/cbork-utils/src/map.rs

```diff
@@ -0,0 +1,605 @@
+    /// the key values."
+    #[test]
+    fn test_map_complex_keys() {
+        // Test nested structures in keys while maintaining order
+        // Following RFC 8949 Section 4.2.3 length-first rule
+        let valid_complex = vec![
+            0xA2, // Map with 2 pairs
+            0x42, 0x01, 0x02, // Key 1: simple 2-byte string (shorter, so comes first)
+            0x41, 0x01, // Value 1
+            0x44, 0x01, 0x02, 0x03, 0x04, // Key 2: 4-byte string (longer, so comes second)
+            0x41, 0x02, // Value 2
+        ];
+        let mut decoder = Decoder::new(&valid_complex);
+        assert!(Map::decode(&mut decoder, &mut DecodeCtx::Deterministic).is_ok());
+    }
+
+    /// Test edge cases for map encoding while maintaining compliance with RFC 8949
+    ///
+    /// These cases test boundary conditions that must still follow all rules from
+    /// Section 4.2:
+    /// - Minimal length encoding (4.2.1)
+    /// - No indefinite lengths (4.2.2)
+    /// - Canonical ordering (4.2.3)
+    #[test]
+    fn test_map_edge_cases() {
+        // Single entry map - must still follow minimal length encoding rules
+        let single_entry = vec![
+            0xA1, // Map with 1 pair (using immediate value as per Section 4.2.1)
+            0x41, 0x01, // Key: 1-byte string
+            0x41, 0x02, // Value: 1-byte string
+        ];
+        let mut decoder = Decoder::new(&single_entry);
+        assert!(Map::decode(&mut decoder, &mut DecodeCtx::Deterministic).is_ok());
+
+        // Map with zero-length string key - tests smallest possible key case
+        // Still must follow sorting rules from Section 4.2.3
+        let zero_length_key = vec![
+            0xA1, // Map with 1 pair
+            0x40, // Key: 0-byte string (smallest possible key length)
+            0x41, 0x01, // Value: 1-byte string
+        ];
+        let mut decoder = Decoder::new(&zero_length_key);
+        assert!(Map::decode(&mut decoder, &mut DecodeCtx::Deterministic).is_ok());
+    }
+
+    /// Test duplicate key detection as required by RFC 8949 Section 4.2.3
+    ///
+    /// From RFC 8949 Section 4.2.3:
+    /// "The keys in every map must be sorted [...] Note that this rule
+    /// automatically implies that no two keys in a map can be equal (have
+    /// the same length and the same value)."
+    #[test]
+    fn test_duplicate_keys() {
+        let map_with_duplicates = vec![
+            0xA2, // Map with 2 pairs
+            0x41, 0x01, // Key 1: 1-byte string [0x01]
+            0x41, 0x02, // Value 1
+            0x41, 0x01, // Key 2: same as Key 1 (duplicate - invalid)
+            0x41, 0x03, // Value 2
+        ];
+        let mut decoder = Decoder::new(&map_with_duplicates);
+        assert!(Map::decode(&mut decoder.clone(), &mut DecodeCtx::Deterministic).is_err());
+        assert!(Map::decode(&mut decoder, &mut DecodeCtx::non_deterministic()).is_ok());
+    }
+
+    #[test]
+    fn test_map_entry_ord_comprehensive() {
```

Comment:
We might consider refactoring some of the tests to use the `test_case` crate. **It's definitely not for this PR**, but let me drop a small example of how this crate can improve test readability: [commit](https://github.com/input-output-hk/catalyst-libs/commit/78044648382dfee18fff3c52f89c9e5826ab3083).